p2p/net/upgrader/multistream.go

```diff
@@ -14,6 +12,17 @@ import (
 
 var DefaultNegotiateTimeout = time.Second * 60
 
+type Multiplexer struct {
+	ID          string
+	StreamMuxer network.Multiplexer
+}
+
+type MsTransport interface {
```

Comment:
Generally callers define the interface: https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_let_callers_define_the_interface_they_require.

There's a tendency to follow Java patterns and say "here's an interface that clients can use". But that's backwards, it's really here's what I provide as my concrete type, and callers can program against the interface they need. If your type doesn't conform to what they need they'll get a compiler error. More context here too: https://github.com/libp2p/go-libp2p/pull/1758#issuecomment-1267623184

Also see https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#interfaces

Reply:
MsTransport removed, thanks for the info link!